rfcs/20191212-keras-categorical-inputs.md

```diff
@@ -0,0 +1,390 @@
+             combined into all combinations of output with degree of `depth`. For example,
+             with inputs `a`, `b` and `c`, `depth=2` means the output will be [ab;ac;bc]
+      num_bins: Number of hash bins. By default None, no hashing is performed.
+      name: Name to give to the layer.
+      **kwargs: Keyword arguments to construct a layer.
+
+    Input: a list of int tensors of shape `[batch_size, d1, ..., dm]`
+    Output: a single int tensor of shape `[batch_size, d1, ..., dm]`
+
+    Example:
+
+    If the layer receives two inputs, `a=[[1, 2]]` and `b=[[1, 3]]`,
+    and if depth is 2, then
+    the output will be a single integer tensor `[[i, j, k, l]]`, where:
+    i is the index of the category "a1=1 and b1=1"
+    j is the index of the category "a1=1 and b2=3"
+    k is the index of the category "a2=2 and b1=1"
+    l is the index of the category "a2=2 and b2=3"
+    """
+    pass
+
+`tf.keras.layers.CategoryEncoding`
+CategoryEncoding(PreprocessingLayer):
+"""This layer transforms categorical inputs from index space to category space.
+   If input is dense/sparse, then output is dense/sparse."""
+
+  def __init__(self, num_categories, mode="sum", axis=-1, name=None, **kwargs):
+    """Constructs a CategoryEncoding layer.
+    Args:
+      num_categories: Number of elements in the vocabulary.
+      mode: how to reduce a categorical input if multivalent, can be one of "sum",  
+          "mean", "binary", "tfidf". It can also be None if this is not a multivalent input,
+          and simply needs to convert input from index space to category space. "tfidf" is only
+          valid when adapt is called on this layer.
+      axis: the axis to reduce, by default will be the last axis, specially true 
+          for sequential feature columns.
+      name: Name to give to the layer.
+     **kwargs: Keyword arguments to construct a layer.
+
+    Input: a int tensor of shape `[batch_size, d1, ..., dm-1, dm]`
+    Output: a float tensor of shape `[batch_size, d1, ..., dm-1, num_categories]`
+    """
+    pass
+
+`tf.keras.layers.CategoryHashing`
+CategoryHashing(PreprocessingLayer):
+"""This layer transforms categorical inputs to hashed output.
+   If input is dense/sparse, then output is dense/sparse."""
+  def __init__(self, num_bins, name=None, **kwargs):
+    """Constructs a CategoryHashing layer.
+
+    Args:
+      num_bins: Number of hash bins.
+      name: Name to give to the layer.
+      **kwargs: Keyword arguments to construct a layer.
+
+    Input: a int tensor of shape `[batch_size, d1, ..., dm]`
+    Output: a int tensor of shape `[batch_size, d1, ..., dm]`
+    """
+    pass
+
+```
+
+We also propose a `to_sparse` op to convert dense tensors to sparse tensors given user specified ignore values. This op can be used in both `tf.data` or [TF Transform](https://www.tensorflow.org/tfx/transform/get_started). In previous feature column world, "" is ignored for dense string input and -1 is ignored for dense int input.
```

Comment:
I would prefer if the `SparseTensor` class had a `from_dense` method.

Reply:
If we don't need the functionality of `sparse_output = to_sparse(sparse_input)`, then `from_dense` is probably better.
This "imagined" functionality is not used anywhere though. In TFT I think any tf.io.VarLenFeature should automatically be sparse input, we just need to call SparseTensor.from_dense for any tf.io.FixedLenFeature.

WDYT?

Reply:
I realized, we already have [from_dense](https://www.tensorflow.org/api_docs/python/tf/sparse/from_dense?version=stable), so perhaps you should just extend it with the option to set the element to be ignore?

Reply:
Yeah good point. I wasn't aware of this op. We should just extend it. Done.